sw/device/silicon_creator/rom_ext/rom_ext_start.S

```diff
@@ -150,12 +159,78 @@ _rom_ext_start_boot:
   la   sp, (_stack_end - 16)
 
   /**
-   * Set well-defined interrupt/exception handlers
-   *
-   * The lowest two bits should be `0b01` to ensure we use vectored interrupts.
+   * Initialize the Ibex simple address translation to run position dependent
+   * code on the virtual flash space.
    */
-  la   t0, (_rom_ext_interrupt_vector + 1)
-  csrw mtvec, t0
+  /**
+   * Compute the current slot address by: 
+   * addr = pc - _addr_translation_crt_offset
+   */
+  la t0, _addr_translation_crt_offset 
+  sub a0, a0, t0 //first address physical
+  // mask = virtual_addr | ((size - 1) >> 1);
+  la a1, _vflash_address
+  la t0, _vflash_size
+  addi t0, t0, -1
+  srli t0, t0, 1
+  or a1, a1,t0 
+
+  // Load ibex config base address
+  li t0, TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR 
+
+  // Set ibus and dbus address matching 
+  sw a1, RV_CORE_IBEX_IBUS_ADDR_MATCHING_0_REG_OFFSET(t0)
+  sw a1, RV_CORE_IBEX_DBUS_ADDR_MATCHING_0_REG_OFFSET(t0)
+
+  // Set ibus and dbus address remap
+  sw a0, RV_CORE_IBEX_IBUS_REMAP_ADDR_0_REG_OFFSET(t0)
+  sw a0, RV_CORE_IBEX_DBUS_REMAP_ADDR_0_REG_OFFSET(t0)
+
+  // Enable address translation
+  li t1, 0x01
+  sw t1, RV_CORE_IBEX_IBUS_ADDR_EN_0_REG_OFFSET(t0)
+  sw t1, RV_CORE_IBEX_DBUS_ADDR_EN_0_REG_OFFSET(t0) 
+ 
+  /**
+   * Configure the epmp to enable execution in the text space only and
+   * read in the whole vFlash space
+   *   | Entry | Description                   | Permissions | Addressing Mode |
+   *   |-------|-------------------------------|-------------|-----------------|
+   *   | 7     | vflash `ROM_EXT` TEXT start   | RX          | TOR             |
+   *   | 8     | vflash `ROM_EXT` end          | RX          | TOR             | 
+   *   | 9     | vFlash                        | R           | NAPOT           |
+   */
+  #define EPMP_CFG_PMPCFG7_OFFSET ((7 * 8) % 32) 
+  #define EPMP_CFG_PMPCFG8_OFFSET ((8 * 8) % 32) 
+  #define EPMP_CFG_PMPCFG9_OFFSET ((9 * 8) % 32) 
+  
+  //pmpaddr7 = (_text_start >> 2) 
+  la t0, _text_start 
+  srli t0, t0, 2
+  csrw pmpaddr7,t0
+  
+  //pmpaddr8 = (_text_end >> 2) 
+  la t0, _text_end
+  srli t0, t0, 2
+  csrw pmpaddr8,t0
+  
+  //Set the bits for the 8th epmp config entry
```

Comment:
It might be worth adding a defensive check that pmpcfg7, pmpcfg8 and pmpcfg9 are all zero at this point. Just in case the mask ROM is changed or this is jumped to from the wrong place. In the event of an error I suggest just executing an `unimp` instruction to trigger an exception.

Reply:
done!